src/kmeans.jl

```diff
@@ -18,35 +18,33 @@ const _kmeans_default_maxiter = 100
 const _kmeans_default_tol = 1.0e-6
 const _kmeans_default_display = :none
 
-function kmeans!(X::Matrix{T}, centers::Matrix{T};
-                 weights=nothing,
-                 maxiter::Integer=_kmeans_default_maxiter,
+function kmeans!(X::AbstractMatrix{<:Real}, centers::AbstractMatrix{<:Real};
+                 weights=nothing, maxiter::Integer=_kmeans_default_maxiter,
                  tol::Real=_kmeans_default_tol,
                  display::Symbol=_kmeans_default_display,
-                 distance::SemiMetric=SqEuclidean()) where T<:AbstractFloat
+                 distance::SemiMetric=SqEuclidean())
 
     m, n = size(X)
     m2, k = size(centers)
     m == m2 || throw(DimensionMismatch("Inconsistent array dimensions."))
     (2 <= k < n) || error("k must have 2 <= k < n.")
 
-    assignments = zeros(Int, n)
-    costs = zeros(T, n)
+    assignments = Vector{Int}(undef, n)
+    costs = Vector{Float64}(undef, n)
```

Comment:
Should use `eltype(X)` here too: using the same eltype as `X` can improve performance when using e.g. `Float32` and can help preserving that type in multiplications. Same for the definition of `KMeansResult` above.